### PR TITLE
Improve performance when listing columns in Iceberg

### LINKED ISSUE #23468: Performance Issue with Retrieving Columns from Iceberg Catalog
Hi,

We are experiencing a performance issue when retrieving column metadata from the Iceberg catalog using the following query:

```sql
SELECT * FROM iceberg.information_schema.columns;
```

In our environment this query takes approximately 30 seconds to return results with clean cache, and we have around 500 tables with 3.5k columns in total.

From our analysis of the logs on both the Iceberg catalog and Trino side, it appears that requests to the catalog might be executed sequentially for each table.

### Setup Details:
- **Trino version**: 453

We tried both catalogs:
- **Nessie version**: 0.94.4 -with PostgreSQL configuration
- **Catalog**: [Tabular REST Catalog example (based on JDBC)](https://github.com/tabular-io/iceberg-rest-image) 

[link](https://trinodb.slack.com/archives/CJ6UC075E/p1726594499184929)  to slack discussion

To investigate the performance issue, I set up a Docker Compose environment with  Trino 451 and the latest versions of the REST catalog, and the Nessie catalog (using in-memory configuration), along with MinIO for Iceberg data storage and Jaeger for tracing. This test was conducted on my laptop, which has 32GB of RAM and an Intel Core 9 processor.

I executed the following SQL query for both the REST and Nessie catalogs:

```sql
SELECT * FROM iceberg.information_schema.columns;
```

After each query, I restarted Trino to clear the cache.

Here are the results for 500 tables, each containing 10 columns (a total of 5,000 columns):
- Nessie catalog: 9.09 seconds
- REST catalog: 9.03 seconds

I suspect that the reason I observed these better performance results compared to my Kubernetes cluster is due to the local setup. Running MinIO locally allows for a closer proximity to Trino, reducing latency.

Is there any optimization or configuration change we could apply to improve the performance of this query? Additionally, any insights into the sequential request behavior would be helpful.

Thanks in advance for your assistance!

Here are some of the tracing(thanks @mosiac1 for the advice it is very useful )  that i got from open telemetry:

Timelines of both nessie and rest:
nessie_timeline
rest_timeline

You can see in the nessie(for some reson it is more detailed than the rest) timeline that for each table it makes a get from the s3 and it do it one after the other. 
nessie_get_column_list_timeline

Here is the tracing as table:
nessie_table
rest_table

Here are the jsons of the tracing if someone wants to load them to jaeger:
[rest_jaeger_columns_benchmark.json](https://github.com/user-attachments/files/17067836/rest_jaeger_columns_benchmark.json)
[nessie_jaeger_columns_benchmark.json.json](https://github.com/user-attachments/files/17067842/nessie_jaeger_columns_benchmark.json.json)

### PR DESCRIPTION
**Pull request by shohamyamin**
# Improve Performance for Retrieving Tables Metadata from Iceberg Catalog

## Overview

This PR addresses a performance issue encountered when querying column metadata from the Iceberg catalog in Trino (Issue #[23468](https://github.com/trinodb/trino/issues/23468)). The primary concern is that requests to retrieve table metadata are executed sequentially for each table, which significantly impacts query performance when dealing with a large number of tables.

## Description

The issue manifests when executing the following query on the Iceberg catalog:

```sql
SELECT * FROM iceberg.information_schema.columns;
```

In environments with a large number of tables, the query’s response time increases considerably due to the sequential execution of catalog requests.

### Changes Made

To improve the performance, this PR introduces parallelism in the metadata retrieval process. The old code executed catalog requests sequentially for each table, which resulted in longer execution times. The updated code utilizes `CompletableFuture` to handle requests asynchronously, thereby reducing the overall execution time.

The updated implementation leverages `CompletableFuture` to process tables concurrently, significantly reducing the latency in retrieving column metadata.

## Additional Context and Related Issues

Fixes #23468

### Test Setup & Observations:
 I checked the code changes and get faster query in over x4 faster then without the proposed change
- Trino Version: 463 and trino with the code in master branch with the proposed change lets called it trino 464-SNAPSHOT
- Nessie Version: 0.94.4 (In memory)
- REST Catalog based on JDBC (e.g., Tabular REST Catalog)
- Testing environment: Docker Compose with MinIO for Iceberg storage and Jaeger for tracing.
- Hardware: Intel Core i9 processor, 32GB RAM

### Performance Comparison

As you can see in this table we get around 70% reduce in time for retrieving the columns

| **Number of Tables** | **Trino Version** | **Catalog Setup** | **Query Execution Time (s)** | **Performance Increase (%)** |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **10**                | Trino 463         | REST              | 0.691                         |                              |
|                       |                   | Nessie            | 1.07                          |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 0.677                         |  2.02%    |
|                       |                   | Nessie            | 0.982                         | 8.23%       |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **100**               | Trino 463         | REST              | 1.24                          |                              |
|                       |                   | Nessie            | 1.89                          |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 0.772                         | 37.90%      |
|                       |                   | Nessie            | 1.29                          | 31.59%       |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **1,000**             | Trino 463         | REST              | 3.06                          |                              |
|                       |                   | Nessie            | 6.55                          |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 1.17                          | 61.59%      |
|                       |                   | Nessie            | 2.42                          | 63.01%      |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **10,000**            | Trino 463         | REST              | 11.38                         |                              |
|                       |                   | Nessie            | 29.59                         |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 3.41                          | 70.04%    |
|                       |                   | Nessie            | 7.70                          | 73.00%     |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **100,000**             | Trino 463         | REST              | 105                           |                              |
|                       |                   | Nessie            | 255                           |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 16.96                         | 83.86%        |
|                       |                   | Nessie            | 34.72                         | 86.38%        |

- Each table in this  comparison has 10 columns
- I ran each test several times, and the results were largely consistent.

## Release Notes

```md
## Iceberg
* Improve performance when listing columns. 
```

### CODE REVIEW

**Comment by alonahmias** [comment]
Thank you for this PR! This solution makes a huge difference in reducing the time needed to retrieve all available columns in Iceberg. The performance improvement is very noticeable, especially with larger datasets

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -865,23 +867,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
...
-                        catch (UnknownTableTypeException e) {
-                            // Skip unsupported table type in case that the table redirects are not enabled
-                        }
-                        catch (RuntimeException e) {
-                            // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
-                            log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
-                        }
-                    }
+                    List<CompletableFuture<Void>> futures = remainingTables.stream()
+                            .map(tableName -> CompletableFuture.runAsync(() -> {
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:871]: We generally don't use the common fork join pool in Trino. We should inject a dedicated executor which is bounded at some number and use that.
GET_METADATA_BATCH_SIZE is 1000, I don't think we want that many parallel fetches, something like 8 parallel fetches is probably enough.

Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:891]: See the way BoundedExecutor is setup in https://github.com/trinodb/trino/pull/24110/files#diff-e72d63b77858985902ad4bb32d26e11cfaab0470c78821ce5416c2c97da4deecR222

Comment (shohamyamin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:904]: @raunaqmorarka Yes I was exploring how this was implemented there earlier. I’ve just updated the code. Could you please review it again? This time, I’ve incorporated the `metadataParallelism` configuration parameter and switched to using a `FixedThreadPool` for managing threads.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java [hunk 2]
```diff
@@ -79,6 +81,6 @@ public IcebergMetadata create(ConnectorIdentity identity)
                 tableStatisticsWriter,
                 metastoreFactory,
                 addFilesProcedureEnabled,
-                allowedExtraProperties);
+                allowedExtraProperties, metadataParallelism);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java:84]: Put `metadataParallelism` on a new line.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java [hunk 3]
```diff
@@ -120,7 +120,7 @@ public void testNonLowercaseNamespace()
                     new TableStatisticsWriter(new NodeVersion("test-version")),
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:123]: Put `8` on a new line.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestTrinoGlueCatalog.java [hunk 4]
```diff
@@ -139,7 +139,7 @@ public void testNonLowercaseGlueDatabase()
                     new TableStatisticsWriter(new NodeVersion("test-version")),
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestTrinoGlueCatalog.java:142]: Put `8` on a new line.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/nessie/TestTrinoNessieCatalog.java [hunk 5]
```diff
@@ -189,7 +189,7 @@ public void testNonLowercaseNamespace()
                     new TableStatisticsWriter(new NodeVersion("test-version")),
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/nessie/TestTrinoNessieCatalog.java:192]: Put `8` on a new line.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/rest/TestTrinoRestCatalog.java [hunk 6]
```diff
@@ -127,7 +127,7 @@ public void testNonLowercaseNamespace()
                     new TableStatisticsWriter(new NodeVersion("test-version")),
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/rest/TestTrinoRestCatalog.java:130]: Put `8` on a new line.

File: plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java [hunk 7]
```diff
@@ -105,15 +105,15 @@ private static void testTableSetup()
...
-                         BASE_LOCATION = '%s/'""".formatted(TpchTable.NATION.getTableName(), SNOWFLAKE_S3_EXTERNAL_VOLUME, TpchTable.NATION.getTableName()));
+                                CREATE OR REPLACE ICEBERG TABLE %s (
+                                	NATIONKEY NUMBER(38,0),
+                                	NAME STRING,
+                                	REGIONKEY NUMBER(38,0),
+                                	COMMENT STRING
+                                )
+                                 EXTERNAL_VOLUME = '%s'
+                                 CATALOG = 'SNOWFLAKE'
+                                 BASE_LOCATION = '%s/'""".formatted(TpchTable.NATION.getTableName(), SNOWFLAKE_S3_EXTERNAL_VOLUME, TpchTable.NATION.getTableName()));
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java:116]: Revert unrelated change.

File: plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java [hunk 8]
```diff
@@ -122,14 +122,14 @@ REGIONKEY NUMBER(38,0),
...
-                         CATALOG = 'SNOWFLAKE'
-                         BASE_LOCATION = '%s/'""".formatted(TpchTable.REGION.getTableName(), SNOWFLAKE_S3_EXTERNAL_VOLUME, TpchTable.REGION.getTableName()));
+                                CREATE OR REPLACE ICEBERG TABLE %s (
+                                	REGIONKEY NUMBER(38,0),
+                                	NAME STRING,
+                                	COMMENT STRING
+                                )
+                                 EXTERNAL_VOLUME = '%s'
+                                 CATALOG = 'SNOWFLAKE'
+                                 BASE_LOCATION = '%s/'""".formatted(TpchTable.REGION.getTableName(), SNOWFLAKE_S3_EXTERNAL_VOLUME, TpchTable.REGION.getTableName()));
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java:132]: Put `8` on a new line.

File: plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java [hunk 9]
```diff
@@ -223,7 +223,7 @@ public void testNonLowercaseNamespace()
                 new TableStatisticsWriter(new NodeVersion("test-version")),
                 Optional.empty(),
                 false,
-                _ -> false);
+                _ -> false, 8);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java:226]: Put `8` on a new line.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 10]
```diff
@@ -1177,8 +1203,7 @@ public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Con
         TrinoFileSystem fileSystem = fileSystemFactory.create(session.getIdentity(), transaction.table().io().properties());
         try {
             if (!replace && fileSystem.listFiles(location).hasNext()) {
-                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("" +
-                        "Cannot create a table on a non-empty location: %s, set 'iceberg.unique-table-location=true' in your Iceberg catalog properties " +
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("Cannot create a table on a non-empty location: %s, set 'iceberg.unique-table-location=true' in your Iceberg catalog properties " +
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1206]: Revert unrelated change.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 11]
```diff
@@ -430,6 +439,16 @@ public IcebergMetadata(
...
         this.allowedExtraProperties = requireNonNull(allowedExtraProperties, "allowedExtraProperties is null");
+
+        if (metadataParallelism == 1) {
+            this.executorService = directExecutor();
+        }
+        else {
+            this.executorService = Executors.newFixedThreadPool(metadataParallelism, new ThreadFactoryBuilder()
+                    .setNameFormat("iceberg-metadata-pool-%d")
+                    .setDaemon(true) // Optional: Makes threads daemon threads.
+                    .build());
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:450]: Why do we initialize `Executor` in `IcebergMetadata` instead of providing it in `IcebergModule`?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java [hunk 12]
```diff
@@ -120,7 +120,8 @@ public void testNonLowercaseNamespace()
                     new TableStatisticsWriter(new NodeVersion("test-version")),
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false,
+                    8);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:124]: Use `new IcebergConfig().getMetadataParallelism()`. Same for others.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 11 above)
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:450]: Why is the logic different from another metadata executor for listing tables?

Comment (shohamyamin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:450]: I was follow this where in recent or they changed it to a similar thing https://github.com/trinodb/trino/blob/4718011a39f700ab3c8dbe66747c37a9398922f8/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadataFactory.java#L111

Can you give me an example for another metadata executer so I will follow the same logic

And the use of ThreadFactoryBuilder is required because of the maven checks tests that require that each thread will have a name format

Comment (shohamyamin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java]: I have just moved that to `IcebergMetadataFactory` and use the already existing `createIcebergMetadataExecutor` that is located in `IcebergModule`

**Comment by shohamyamin** [comment]
@ebyhr I have change the code to follow the way that this kind of things are done I was taking inspiration from that  [PR](https://github.com/trinodb/trino/pull/24110) that is very similar and I use the already existing `createIcebergMetadataExecutor` that is located in `IcebergModule`

**Comment by ebyhr** [comment]
Could you rebase on master to resolve conflicts?

**Comment by shohamyamin** [comment]
> Could you rebase on master to resolve conflicts?

@ebyhr I’ve rebased as requested, and the PR is now updated and ready for review again. Let me know if there’s anything else I should address.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 13]
```diff
@@ -423,7 +428,8 @@ public IcebergMetadata(
             Optional<HiveMetastoreFactory> metastoreFactory,
             boolean addFilesProcedureEnabled,
             Predicate<String> allowedExtraProperties,
-            ExecutorService executor)
+            ExecutorService executor,
+            Executor metadataFetchingExecutor)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:432]: Now `executor` is too general. Please rename to `systemTableExecutor` or something.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java [hunk 14]
```diff
@@ -53,7 +57,8 @@ public IcebergMetadataFactory(
             TableStatisticsWriter tableStatisticsWriter,
             @RawHiveMetastoreFactory Optional<HiveMetastoreFactory> metastoreFactory,
             @ForIcebergScanPlanning ExecutorService executor,
-            IcebergConfig config)
+            IcebergConfig config,
+            @ForIcebergMetadata ExecutorService metadataExecutorService)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java:61]: Replace the order between `config` and `metadataExecutorService`.

**Comment by ebyhr** [comment]
> Improve Performance for Retrieving Tables Metadata from Iceberg Catalog

You can change the commit title to "Improve performance when listing columns in Iceberg".

**Comment by ebyhr** [comment]
/test-with-secrets sha=971fa03cb8d4113057825347477eee3cabe06586

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 15]
```diff
@@ -981,23 +987,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
...
-                        }
-                    }
+                    List<CompletableFuture<Void>> futures = remainingTables.stream()
+                            .map(tableName -> CompletableFuture.runAsync(() -> {
+                                try {
+                                    Table icebergTable = catalog.loadTable(session, tableName);
+                                    List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema(), typeManager);
+                                    synchronized (tableMetadatas) {
+                                        tableMetadatas.add(TableColumnsMetadata.forTable(tableName, columns));
+                                    }
```
Comment (piotrrzysko) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:996]: I think this synchronized block can be avoided by returning a result (`TableColumnsMetadata`) from each task.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 16]
```diff
@@ -981,23 +986,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
...
+                                    // Skip unsupported table type in case that the table redirects are not enabled
+                                }
+                                catch (RuntimeException e) {
+                                    // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
+                                    log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
+                                }
+                            }, metadataFetchingExecutor))
+                            .toList();
+
+                    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
```
Comment (piotrrzysko) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1011]: How will this behave if the query is cancelled? Will pending tasks also be cancelled? I’d suggest using `ExecutorUtil#processWithAdditionalThreads` instead, as it handles such scenarios.

Comment (shohamyamin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java]: I have updated the code to use`processWithAdditionalThreads `

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java [hunk 17]
```diff
@@ -52,7 +52,7 @@ public IcebergMetadataFactory(
             IcebergFileSystemFactory fileSystemFactory,
             TableStatisticsWriter tableStatisticsWriter,
             @RawHiveMetastoreFactory Optional<HiveMetastoreFactory> metastoreFactory,
-            @ForIcebergScanPlanning ExecutorService executor,
+            @ForIcebergScanPlanning ExecutorService systemTableExecutor,
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java:59]: This executor is not specific to system tables, it is used for planning any iceberg scan which could be system table or not.
Either we shouldn't rename it or rename to icebergScanExecutor

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 18]
```diff
@@ -981,23 +989,40 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
...
+                                    // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
+                                    log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
+                                    return null;
+                                }
+                            })
+                            .collect(toImmutableList());
+
+                    try {
+                        List<TableColumnsMetadata> taskResults = processWithAdditionalThreads(tasks, metadataFetchingExecutor).stream()
+                                .filter(Objects::nonNull) // Filter out null results
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1017]: Use `Optional#empty` instead of `null`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java (hunk 17 above)
Comment (shohamyamin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java]: I've renamed it to `icebergScanExecutor` as you suggested. Keeping it as just `executor` could cause confusion, especially since I'm adding a new executor named `metadataFetchingExecutor`.

**Comment by shohamyamin** [comment]
@raunaqmorarka I update the code as requested

**Review by raunaqmorarka — APPROVED** [review]
